File: backend/internal/middleware/audit.go
package middleware

import (
	"bytes"
	"context"
	"encoding/json"
	"io"
	"log"
	"net/http"
	"strings"
	"time"

	"panel-backend/internal/services"
)

// sensitiveKeys are redacted from audit log body
var sensitiveKeys = map[string]bool{
	"password":         true,
	"token":            true,
	"secret":           true,
	"jwt":              true,
	"current_password": true,
	"new_password":     true,
}

// statusRecorder wraps ResponseWriter to capture status code
type statusRecorder struct {
	http.ResponseWriter
	statusCode int
}

func (r *statusRecorder) WriteHeader(code int) {
	r.statusCode = code
	r.ResponseWriter.WriteHeader(code)
}

// Audit middleware logs POST/PUT/DELETE/PATCH requests
func Audit(db *services.DB) func(http.Handler) http.Handler {
	return func(next http.Handler) http.Handler {
		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
			method := r.Method

			// Only audit mutating methods
			if method != "POST" && method != "PUT" && method != "DELETE" && method != "PATCH" {
				next.ServeHTTP(w, r)
				return
			}

			start := time.Now()

			// Read and restore body for auditing
			var bodyJSON string
			if r.Body != nil && r.Header.Get("Content-Type") == "application/json" {
				bodyBytes, err := io.ReadAll(io.LimitReader(r.Body, 1024*1024)) // 1MB max
				r.Body.Close()
				if err == nil {
					r.Body = io.NopCloser(bytes.NewReader(bodyBytes))
					bodyJSON = sanitizeBody(bodyBytes)
				} else {
					r.Body = io.NopCloser(bytes.NewReader(nil))
					bodyJSON = "{}"
				}
			} else {
				bodyJSON = "{}"
			}

			// Wrap response writer to capture status code
			rec := &statusRecorder{ResponseWriter: w, statusCode: 200}

			next.ServeHTTP(rec, r)

			// Fire-and-forget audit log
			durationMs := int(time.Since(start).Milliseconds())
			username := GetUsername(r)
			ip := extractIP(r)

			go func() {
				ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
				defer cancel()

				_, err := db.Exec(ctx,
					`INSERT INTO audit_log (username, ip, method, path, status_code, duration_ms, body)
					 VALUES ($1, $2, $3, $4, $5, $6, $7)`,
					username, ip, method, r.URL.Path, rec.statusCode, durationMs, bodyJSON,
				)
				if err != nil {
					log.Printf("audit log error: %v", err)
				}
			}()
		})
	}
}

// sanitizeBody redacts sensitive fields from JSON body
func sanitizeBody(raw []byte) string {
	var obj map[string]interface{}
	if err := json.Unmarshal(raw, &obj); err != nil {
		return "{}"
	}

	for key := range obj {
		if sensitiveKeys[strings.ToLower(key)] {
			obj[key] = "[REDACTED]"
		}
	}

	sanitized, err := json.Marshal(obj)
	if err != nil {
		return "{}"
	}
	return string(sanitized)
}

// extractIP gets the client IP from headers or connection
func extractIP(r *http.Request) string {
	// X-Real-IP (set by NGINX)
	if ip := r.Header.Get("X-Real-IP"); ip != "" {
		return ip
	}

	// X-Forwarded-For (first entry)
	if xff := r.Header.Get("X-Forwarded-For"); xff != "" {
		parts := strings.SplitN(xff, ",", 2)
		return strings.TrimSpace(parts[0])
	}

	// RemoteAddr (strip port)
	addr := r.RemoteAddr
	if idx := strings.LastIndex(addr, ":"); idx != -1 {
		return addr[:idx]
	}
	return addr
}

File: backend/src/middleware/audit.ts
import { Request, Response, NextFunction } from 'express';
import { query } from '../services/db';
import { logger } from '../services/logger';

/**
 * Audit logging middleware.
 * Logs all state-changing requests (POST, PUT, DELETE) to the audit_log table.
 * GET requests are not logged to avoid noise.
 */
export function auditMiddleware(req: Request, res: Response, next: NextFunction): void {
  // Only log state-changing operations
  if (!['POST', 'PUT', 'DELETE', 'PATCH'].includes(req.method)) {
    next();
    return;
  }

  // Capture original end to log after response is sent
  const originalEnd = res.end;
  const startTime = Date.now();

  // Override res.end to capture status code
  res.end = function (...args: Parameters<typeof originalEnd>) {
    const duration = Date.now() - startTime;
    const username = req.user?.sub ?? 'anonymous';
    const ip =
      (req.headers['x-real-ip'] as string) ??
      (req.headers['x-forwarded-for'] as string)?.split(',')[0]?.trim() ??
      req.ip ??
      'unknown';

    // Sanitize body — never log passwords or tokens
    const sanitizedBody = { ...req.body };
    for (const key of ['password', 'token', 'secret', 'jwt', 'current_password', 'new_password']) {
      if (key in sanitizedBody) sanitizedBody[key] = '***REDACTED***';
    }

    // Fire-and-forget — don't block the response
    query(
      `INSERT INTO audit_log (username, ip, method, path, status_code, duration_ms, body)
       VALUES ($1, $2, $3, $4, $5, $6, $7)`,
      [
        username,
        ip,
        req.method,
        req.originalUrl,
        res.statusCode,
        duration,
        JSON.stringify(sanitizedBody),
      ]
    ).catch((err) => {
      logger.error('Failed to write audit log', { error: err.message });
    });

    return originalEnd.apply(res, args);
  } as typeof originalEnd;

  next();
}
